### ISSUE TITLE
.NET 7 Support?

### ISSUE THREAD

**Issue by cdomino** [Tewr/BlazorWorker#85]
Good Morning!

I've been loving this library, but unfortunately it's no longer working after upgrading to .NET 7.  Here's what I'm getting as soon as I call `await this._backgroundService.CreateInstance`

_dotnet.7.0.0.amub20uvka.js:8 Uncaught (in promise) SyntaxError: Cannot use 'import.meta' outside a module_

...which is this call (line 168) in your JS:

`self.importScripts('${initConf.appRoot}/${initConf.wasmRoot}/${dotnetjsfilename}');`

Please let me know if this is a quick fix or if there is a workaround.

Thanks so much!

**Comment by fgilde** [comment]
Same issue.

**Comment by cdomino** [comment]
Looks like dudes in sunglasses who are quite excited about multi-threaded Blazor WASM are struggling with this... 😄 

[This NuGet package](https://github.com/LostBeard/SpawnDev.BlazorJS) gets past the JS issue above, but I couldn't get it work all the way. I'm going to take a crack at building this myself over the next week, hoping for a Christmas miracle. 

Thanks!

**Comment by fgilde** [comment]
I not really struggeling with it. I have also implemented a Bridge to js webworker, and I was thinking to optimize it amd maybe build a package for it. Then I found this package and because the using is very similar I would have implement it I want to use this instead. However thanks for the Link to the other package. But this has a problem in my opinion. 
It uses reflection in the static JS class what is not the best idea but ok. But more scared makes me the implementation because thats using `Microsoft.AspNetCore.Components.WebAssembly.Hosting` and is not available on ServerSide rendered pages and so on.

**Comment by Serg046** [comment]
Same thing. Found the library just some minutes ago and got into this issue. The lib looks very useful but I need it on .NET 7.

**Comment by Serg046** [comment]
Looking at this similar PR [Net6 Support](https://github.com/Tewr/BlazorWorker/pull/66) and see there might be a need to target .NET 7 as well as .NET 6 that time. And [this one](https://github.com/Tewr/BlazorWorker/commit/bc42773a8b8abdc0224c0e9c069015f60175495d#diff-58c74bb2690a944d5da963dddd152a7addc44616c20226c6611e4cec7cdc4b09) concerns me even more:

![image](https://user-images.githubusercontent.com/4963385/209484007-f34648bd-8d7b-471e-9a86-c14640c38eda.png)

**Comment by Tewr** [comment]
Thank you for the bug report. I'll try to look into this the coming days

**Comment by danielpastoor** [comment]
Is there an update @Tewr

**Comment by Tewr** [comment]
Well I know what the problem is... Referenced framework scripts that used to be classic scripts are now modules and that doesn't play well in a worker... The library mentioned above seems to have solved the problems by rewriting the referenced framework scripts on the fly, replacing imports etc with equivalent classic script constructs. I might end up doing the same. Would be nice to have a cleaner solution though

**Comment by danielpastoor** [comment]
I think Dotnet has it own fix for this problem. Because they are using dotnet.worker.js. This seems to be a compatible script with the dotnet runtime for a worker.

**Comment by Tewr** [comment]
@danielpastoor sounds interesting. Do you have more info on this script in .net7? is it part of the experimental multithreading workload?

**Comment by Tewr** [comment]
Some news: I've tried de-modulizing the dotnet script and while it's technically working, all of the current bootstrap code in BlazorWorker.js is completely out of date, so no initialization (eg call to onReady()) is done. I'm trying to figure of what the new bootstrapper workflow is, but it's not trivial. Might be up for a full rewrite. I'll have to look into [the wasm-experimental workload](https://learn.microsoft.com/en-us/aspnet/core/client-side/dotnet-interop?view=aspnetcore-7.0) to see if there are any hints

**Comment by danielpastoor** [comment]
I also tried rewriting it to a non module vader script within your structure but when I start the script it needs a mono config file. How did you get this working?

**Comment by Tewr** [comment]
I didn't get it working, I've done some debugging and reading though. As I understand it, the mono-config.json is an artifact from the build, the "standard" way of loading dotnet, whereas blazor is overriding the default behaviour, using blazor-boot.json instead (but not in-place, they convey the same information but with different formats). I think understanding how and where this override is done is one of the keys of getting this working.

**Comment by americanslon** [comment]
Any new movements on this front?

**Comment by Tewr** [comment]
Technically, i think this is solved for .net7. But the build process with multitargeting is complex. I haven't found the time these last past weeks to put in the time needed for something maintainable. I might have to split into multiple projects or something, or possibly abandon support for anything lower than .net7 it might speed things up. In any case, i won't be able to get anything out before July.

**Comment by HarryCordewener** [comment]
Any new info on this?

**Comment by Tewr** [comment]
Hello. This has been low on my priority  list this summer and I haven't been very motivated as .net8 promised native multithreading. Turns out it's been postponed (again) to .net9 -- so this might be worth a shot. I'll try to put in the hours needed so that it's ready about the same time as for the release of .net8 (middle of November)

**Comment by krauthaufen** [comment]
Same problem here, any hints on what needs to be done?
I can't really wait until november for the fix but I can invest some time.

Any pointers would be greatly appreciated

**Comment by cdomino** [comment]
Hey @krauthaufen - I took inspiration from @Tewr and [this project](https://github.com/LostBeard/SpawnDev.BlazorJS) (which didn't work for me, but had some useful WASM bootstrapping JS bits) to get something working for .NET 7 (and beyond, as this keeps getting delayed). Let me find time to package it up and I'll report back!

**Comment by Tewr** [comment]
> Same problem here, any hints on what needs to be done? I can't really wait until november for the fix but I can invest some time.
> 
> Any pointers would be greatly appreciated

The branch/pr which is referenced above is a good starting point. I guess what is missing is the generation of a mono-config.json with the correct content, probably by using proxy or something similar so that the standard framework uses it. If you compile one by hand and drop it in the build output, it works more or less. In addition to that, either make multi-targeting work, or removing all multi-targeting code in favor of .net7 and above only target for the code base.

**Comment by krauthaufen** [comment]
> Hey @krauthaufen - I took inspiration from @Tewr and [this project](https://github.com/LostBeard/SpawnDev.BlazorJS) (which didn't work for me, but had some useful WASM bootstrapping JS bits) to get something working for .NET 7 (and beyond, as this keeps getting delayed). Let me find time to package it up and I'll report back!

I'd be glad to see it, I actually don't need all the serialization-, RPC-bits but only want to send Uint8Arrays (byte[]) back and forth, but I have no idea how to setup a dotnet-runtime in the worker. I'll try to figure it out but i guess someone with more experience may be a lot faster here.

**Comment by cdomino** [comment]
Hey @krauthaufen - I'm getting ready to release my own flavor of this after using it extensively on the app I'm building. I will need to surgically isolate it and refactor a few things so it's happy living in Nuget. But by the time that's done, .NET 8 will be upon us, so it's probably best to wait for that to RTM and test with the latest and greatest. I took the week of Thanksgiving off to finish this, so hopefully that gives it at around a year of life before Microsoft (finally/hopefully/please please PLEASE for the love of science) releases proper JavaScript multithreading in Blazor WASM in .NET 9 (<- wow that makes me feel old).

Thanks!

**Comment by krauthaufen** [comment]
Hey, i managed to get that running myself and it's now part of our WASM library [here](https://github.com/aardworx/aardworx.webassembly/blob/master/src/Aardworx.WebAssembly/Worker.fs).

If anyone is interested I could also peel it out and create a separate package for it

**Comment by Nelfstor** [comment]
> If anyone is interested I could also peel it out and create a separate package for it

Sorry, for probably extra-clarifying: do you manage to run this library (by Tewr) with your library?

**Comment by krauthaufen** [comment]
Actually no, i just wrote a very basic thing where the worker and the main thread can exchange strings and byte arrays without all the serialization/rpc stuff on top.

**Comment by Tewr** [comment]
Hello all, 

Some news, I finally got a working prototype in https://github.com/Tewr/BlazorWorker/pull/89 . Had to catch up on many projects and unfortunately this one was on the bottom of my list. The PR still contains some hard-coded strings and stuff, and there seems to be a problem when serializing the response, but I think most of the major problems has been solved.. I'm probably going sacrify/deprecate the current JSRuntime / IJSRuntime plugin, as it seems that in any case the new native code-generator JsImport/JsExport can be a good replacement for that functionality. Also the dependency declarations will probably not be functional , this may mean some additional boot time for worker startup.

 I'm very positive in creating a preview release in the coming days.
